src/mongocxx/test/client_side_encryption.cpp

```diff
@@ -322,28 +360,569 @@ TEST_CASE("Datakey and double encryption", "[client_side_encryption]") {
+    }
+}
+
 TEST_CASE("External key vault", "[client_side_encryption]") {
-    // TODO CXX-2021
+    instance::current();
+
+    class client setup_client {
+        uri {}
+    };
+    if (test_util::get_max_wire_version(setup_client) < 8) {
+        // Automatic encryption requires wire version 8.
+        WARN("Skipping - max wire version is < 8");
+        return;
+    }
+
+    run_external_key_vault_test(true);
+    run_external_key_vault_test(false);
 }
 
 TEST_CASE("BSON size limits and batch splitting", "[client_side_encryption]") {
-    // TODO CXX-2021
+    instance::current();
+
+    // Create a MongoClient without encryption enabled (referred to as client).
+    class client client {
+        uri {}
+    };
+
+    if (test_util::get_max_wire_version(client) < 8) {
+        // Automatic encryption requires wire version 8.
+        WARN("Skipping - max wire version is < 8");
+        return;
+    }
+
+    // Load in json schema limits/limits-schema.json and limits/limits-key.json
+    auto limits_schema = _doc_from_file("/limits/limits-schema.json");
+    auto limits_key = _doc_from_file("/limits/limits-key.json");
+
+    // Using client, drop db.coll and keyvault.datakeys.
+    _setup_drop_collections(client);
+
+    // Create the collection db.coll configured with the JSON schema limits/limits-schema.json.
+    auto db = client["db"];
+    auto cmd = document{} << "create"
+                          << "coll"
+                          << "validator" << open_document << "$jsonSchema" << limits_schema.view()
+                          << close_document << finalize;
+    db.run_command(cmd.view());
+
+    // Insert the document limits/limits-key.json into keyvault.datakeys.
+    auto datakeys = client["keyvault"]["datakeys"];
+    datakeys.insert_one(limits_key.view());
+
+    // Create a MongoClient configured with auto encryption (referred to as client_encrypted),
+    // with local KMS provider as follows:
+    //     { "local": { "key": <base64 decoding of LOCAL_MASTERKEY> } }
+    // and with the keyVaultNamespace set to keyvault.datakeys.
+    char key_storage[96];
+    memcpy(&(key_storage[0]), kLocalMasterKey, 96);
+    bsoncxx::types::b_binary local_master_key{
+        bsoncxx::binary_sub_type::k_binary, 96, (const uint8_t*)&key_storage};
+
+    auto kms_doc = document{} << "local" << open_document << "key" << local_master_key
+                              << close_document << finalize;
+
+    options::client client_encrypted_opts;
+    _add_client_encrypted_opts(&client_encrypted_opts, limits_schema.view(), std::move(kms_doc));
+    class client client_encrypted {
+        uri{}, std::move(client_encrypted_opts)
+    };
+
+    // Using client_encrypted perform the following operations:
+    // Insert { "_id": "over_2mib_under_16mib",
+    //          "unencrypted": <the string "a" repeated 2097152 times> }.
+    std::string over_2mib_under_16mib(static_cast<size_t>(2097152), 'a');
```

Comment:
Yeah it's a handy constructor!  Ah, yes, the static_cast is probably a hold-over from when I was trying to use brace initialization, and the compiler kept trying to cast that number to a char instead of calling this constructor.  Removed.